src/Microsoft.AspNetCore.Razor.Language/LargeTextSourceDocument.cs

```diff
@@ -101,6 +103,27 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
             }
         }
 
+        public override byte[] GetChecksum()
+        {
+            if (_checksum == null)
+            {
+                var charBuffer = new char[Length];
+                CopyTo(0, charBuffer, 0, Length);
+
+                var encoder = Encoding.GetEncoder();
+                var byteCount = encoder.GetByteCount(charBuffer, 0, charBuffer.Length, flush: true);
+                var bytes = new byte[byteCount];
+                encoder.GetBytes(charBuffer, 0, charBuffer.Length, bytes, 0, flush: true);
+
+                using (var hashAlgorithm = SHA1.Create())
+                {
+                    _checksum = hashAlgorithm.ComputeHash(bytes);
+                }
+            }
+
+            return _checksum;
```

Comment:
`byte[]` is mutable. Either don't cache, or make a copy each time you return it.